src/apps/EDGE3D-Cuda.cpp

```diff
@@ -0,0 +1,129 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2017-22, Lawrence Livermore National Security, LLC
+// and RAJA Performance Suite project contributors.
+// See the RAJAPerf/LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include "EDGE3D.hpp"
+
+#include "RAJA/RAJA.hpp"
+
+#if defined(RAJA_ENABLE_CUDA)
+
+#include "common/CudaDataUtils.hpp"
+
+#include "AppsData.hpp"
+
+#include <iostream>
+
+namespace rajaperf
+{
+namespace apps
+{
+
+#define EDGE3D_DATA_SETUP_CUDA \
+  allocAndInitCudaDeviceData(x, m_x, m_array_length); \
+  allocAndInitCudaDeviceData(y, m_y, m_array_length); \
+  allocAndInitCudaDeviceData(z, m_z, m_array_length); \
+  allocAndInitCudaDeviceData(sum, m_sum, m_array_length);
+
+#define EDGE3D_DATA_TEARDOWN_CUDA \
+  getCudaDeviceData(m_sum, sum, m_array_length); \
+  deallocCudaDeviceData(x); \
+  deallocCudaDeviceData(y); \
+  deallocCudaDeviceData(z); \
+  deallocCudaDeviceData(sum);
+
+template < size_t block_size >
+__launch_bounds__(block_size)
+__global__ void edge3d(Real_ptr sum,
+                      const Real_ptr x0, const Real_ptr x1,
+                      const Real_ptr x2, const Real_ptr x3,
+                      const Real_ptr x4, const Real_ptr x5,
+                      const Real_ptr x6, const Real_ptr x7,
+                      const Real_ptr y0, const Real_ptr y1,
+                      const Real_ptr y2, const Real_ptr y3,
+                      const Real_ptr y4, const Real_ptr y5,
+                      const Real_ptr y6, const Real_ptr y7,
+                      const Real_ptr z0, const Real_ptr z1,
+                      const Real_ptr z2, const Real_ptr z3,
+                      const Real_ptr z4, const Real_ptr z5,
+                      const Real_ptr z6, const Real_ptr z7,
+                      Index_type ibegin, Index_type iend)
+{
+   Index_type ii = blockIdx.x * block_size + threadIdx.x;
+   Index_type i = ii + ibegin;
```

Comment:
Why do we need two indexes here? `ii` appears to be unused.

Reply:
i = ii + ibegin ?

Reply:
I think the suggestion here is `Index_type i = ibegin + blockIdx.x * block_size + threadIdx.x;`